### PR TITLE
[OGR provider] changeAttributeValues(): validate field index + FID (fixes #54816)

### LINKED ISSUE #54816: QgsVectorDataProvider.changeAttributeValues() returns True even when it fails
### What is the bug or the crash?

When trying to update a vector layer's feature's attributes using `QgsVectorDataProvider.changeAttributeValues()` method, `False` is returned when an incompatible value is entered (eg string instead of int), but `True` is returned when an invalid `featureId` or `fieldIndex` is passed.

### Steps to reproduce the issue

``` python
l=QgsVectorLayer('output/data/resources/data/world_map.gpkg|layername=countries')

# bad field value correctly returns False
l.dataProvider().changeAttributeValues({1: {0:"asdf"}})  # returns False (correct)

# missing fid always returns True
l.dataProvider().changeAttributeValues({})  # returns True
l.dataProvider().changeAttributeValues({-1: {}})  # returns True
l.dataProvider().changeAttributeValues({-1: {42:42}})  # returns True
l.dataProvider().changeAttributeValues({-1: {'asdf':42}})  # returns True

# existing fid but missing/bad field id returns True
l.dataProvider().changeAttributeValues({1: {}})  # returns True
l.dataProvider().changeAttributeValues({1: {42:42}})  # returns True
l.dataProvider().changeAttributeValues({1: {'asdf':42}})  # returns True
```

### Versions

defb622e, gdal 3.7.1

### Supported QGIS version

- [X] I'm running a supported QGIS version according to [the roadmap](https://www.qgis.org/en/site/getinvolved/development/roadmap.html#release-schedule).

### New profile

- [X] I tried with a new [QGIS profile](https://docs.qgis.org/latest/en/docs/user_manual/introduction/qgis_configuration.html#working-with-user-profiles)

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by rouault**
Also add consistency checks between Qt values and OGR field types in addFeature() and changeAttributeValues()

### CODE REVIEW

File: src/core/providers/ogr/qgsogrprovider.cpp [hunk 1]
```diff
@@ -1574,91 +1610,127 @@ bool QgsOgrProvider::addFeaturePrivate( QgsFeature &f, Flags flags, QgsFeatureId
...
+            const int count = list.count();
+            int *lst = new int[count];
+            if ( count > 0 )
             {
-              lst[pos] = value.toInt();
-              pos++;
+              int pos = 0;
+              for ( const QVariant &value : list )
+              {
+                lst[pos] = value.toInt( &ok );
```
Comment (nyalldawson) [inline src/core/providers/ogr/qgsogrprovider.cpp:1656]: Does this need the same integer truncation test as above?

Comment (rouault) [reply src/core/providers/ogr/qgsogrprovider.cpp]: good point. addressed by extra commit

**Review by nyalldawson — APPROVED** [review]

**Comment by PeterPetrik** [comment]
related new issue to this PR: https://github.com/qgis/QGIS/issues/55517

**Comment by agiudiceandrea** [comment]
Hi @rouault, have you seen the issue report https://github.com/qgis/QGIS/issues/55517? Is it related with this PR?

**Pull request merged**